examples/ctc.dx

```diff
@@ -13,89 +13,123 @@ However, Dex's autodiff can produce the backward
 
 
 '### Main CTC algorithm
-Computes p(labels | logits), marginalizing over possible alignments and
+Computes `p(labels | logits)`, marginalizing over possible alignments and
 and marginalizes over insertion of blanks between characters.
 Todo: add a `[Subset position time]` constraint.
 
-def ctc_inner {vocab time position} [NonEmpty vocab, NonEmpty position, NonEmpty time, Eq time, Eq vocab]
+def ctc_inner {vocab time position}
+  [NonEmpty vocab, NonEmpty position, NonEmpty time, Eq time, Eq vocab]
     (blank: vocab)
     (logits: time=>vocab=>LogSpace Float)
     (labels: position=>vocab)
       : LogSpace Float =
-  ilabels = prepend_and_interleave blank labels
-  normalized_logits = for t. normalize logits.t
+  ilabels = interleave_blanks blank labels
+  label_probs = for t. normalize logits.t
 
-  -- Initialize dynamic programming table.
-  logprob_start_with_blank = normalized_logits.first_ix.blank
-  logprob_of_first_label   = normalized_logits.first_ix.(labels.first_ix)
-  log_prob_seq_t0 = yield_state (for pos. Exp (-10000.0)) \lpRef.
-    lpRef!(Left  first_ix) := logprob_start_with_blank
-    lpRef!(Right first_ix) := logprob_of_first_label  -- 2nd index
+  -- Initialize dynamic programming table to all zeros
+  -- except for starting with either a blank or the first token.
+  prob_seq_at_start = yield_state zero \lpRef.
+    lpRef!(Posts first_ix) := label_probs.first_ix.blank
+    lpRef!(Fence first_ix) := label_probs.first_ix.(labels.first_ix)
 
   safe_idx_sub = \prev s y.
     if (ordinal s) < y
       then zero
       else prev.(unsafe_from_ordinal _ (unsafe_nat_diff (ordinal s) y))
 
-  log_prob_seq_final = fold zero \t prev.
-    case t == first_ix of
-      True -> log_prob_seq_t0
-      False -> for s.
-        same_as_last = if (ordinal s) >= 2
-          then ilabels.s == ilabels.(unsafe_from_ordinal _ ((ordinal s) -| 2))
-          else False
-
-        -- Equation 6 from CTC paper.
-        ans = if ilabels.s == blank || same_as_last
-          then prev.s + safe_idx_sub prev s 1
-          else prev.s + safe_idx_sub prev s 1 + safe_idx_sub prev s 2
-        ans * normalized_logits.t.(ilabels.s)
-
-  -- Sum over last two entries in the dynamic programming table.
-  last_label = log_prob_seq_final.(Right (last_ix, False))
-  last_space = log_prob_seq_final.(Right (last_ix, True ))
-  last_label + last_space
-
+  same_as_last = \s. if (ordinal s) >= 2
+    then
+      s_minus_2 = unsafe_from_ordinal _ (unsafe_nat_diff (ordinal s) 2)
+      ilabels.s == ilabels.s_minus_2
+    else False
+
+  prob_seq_final = fold prob_seq_at_start \t:(AllButFirst time) prev.
+    for s.
+      prob_stay_same       = prev.s
+      prob_from_blank      = safe_idx_sub prev s 1
+      prob_from_last_token = safe_idx_sub prev s 2
+
+      -- Equation 6 from CTC paper.
+      ans = if ilabels.s == blank || same_as_last s
+        then prob_stay_same + prob_from_blank
+        else prob_stay_same + prob_from_blank + prob_from_last_token
```

Comment:
I didn't read the paper, but it looks weird to have `prob_stay_same` appear if the label is not the same?  If that's intended, maybe a brief comment would help?

Reply:
Good point, these are poorly named.  The paper is just a nest of equations without any names.  `same_as_last` and `stay_same` have a confusing collision because there are two distinct notions of time (one for the tokens and one for the sounds).  I'll take another stab at these names.

Reply:
Renamed to hopefully be clearer.